File: Example/graph.php
<?php

/**
 * 使用邻接矩阵 存储如下图 （ps 邻接表是用链表存储）
 * 0    1    2
 * 3    4    5
 *      6    7
 */
$graph       = [];
$graph[0][1] = 1;
$graph[0][3] = 1;
$graph[1][2] = 1;
$graph[1][4] = 1;
$graph[2][1] = 1;
$graph[2][5] = 1;
$graph[3][4] = 1;
$graph[4][1] = 1;
$graph[4][3] = 1;
$graph[4][5] = 1;
$graph[4][6] = 1;
$graph[6][4] = 1;
$graph[6][7] = 1;
$graph[5][2] = 1;
$graph[5][4] = 1;
$graph[5][7] = 1;

// 广度优先寻找最短路径
function bfs($graph, $j, $s): array
{

    $visited = []; //判断节点是否 被访问过
    $queue   = [];
    $path    = []; // 访问路径

    $queue[]     = $j;
    $visited[$j] = true;
    while (!is_null($q = array_shift($queue))) {
        foreach ($graph[$q] as $key => $item) {
            if ($visited[$key]) {
                continue;
            }
            $path[] = $key;
            if ($key == $s) {
                return $path;
            }
            $queue[]       = $key;
            $visited[$key] = true;
        }
    }
    return $path;
}


function dfs($graph, $j, $s): array
{
    $visited = [];
    $path    = [];
    $found = false;

    _dfs($j, $s, $graph, $visited, $path, $found);
    return $path;
}
function _dfs($j, $s, $graph, &$visited, &$path, &$found)
{
    if ($j == $s) {
        $found = true;
    }
    if ($found) {
        return true;
    }
    $visited[$j] = true;

    foreach ($graph[$j] as $key => $item) {
        if (!$visited[$key]) {
            $path[] = $key;
             _dfs($key, $s, $graph, $visited, $path, $found);
        }
    }
}

print_r(dfs($graph, 0, 7));

File: Example/graph.go
package main

import "fmt"

type Queue struct {
	list []int
}
func NewQueue() *Queue{
	list := []int{}
	return &Queue{list:list}
}
func (q *Queue)Push(data int) {
	q.list = append(q.list, data)
}

func (q *Queue)Pop() int{
	if len(q.list)==0 {
		return 0
	}
	res := q.list[0]
	q.list = q.list[1:]
	return res
}


func main()  {

	// 使用邻接矩阵构建无向图
	graph := newGraph()

	fmt.Printf("%c", graph)

	//bfs(graph, 0,7)
}

// 广度优先 从s 到 t 的最短路劲
func bfs(graph [][]int, s int, t int)  {
	var visited []bool
	var queue = NewQueue()
	var prev []int

	visited[s] = true
	queue.Push(s)

	for len(queue.list) > 0{
		w := queue.Pop()
		for i := 0; i < len(graph[w]); i++ {
			q := graph[w][i]
			if !visited[q]{
				prev[w] = q
				if t == q {
					fmt.Printf("%c", prev)
					return
				}
			}
			queue.Push(q)
		}
	}
}

func newGraph() [][]int {
	/**
	0	1	2
	3	4	5
		6	7
	*/
	var graph [][]int
	addNode(0,1,graph)
	addNode(0,3,graph)
	addNode(1,2,graph)
	addNode(1,4,graph)
	addNode(3,4,graph)
	addNode(4,5,graph)
	addNode(4,6,graph)
	addNode(6,7,graph)
	addNode(5,7,graph)
	return graph
}

func addNode(i int, j int, graph [][]int)  {
	graph[i][j] = 1
	graph[j][i] = 1
}